test/comps/test_qb.cc

```diff
@@ -15,62 +15,78 @@ class TestQB : public ::testing::Test
 
     virtual void TearDown() {};
 
+    template <typename T>
+    struct QBTestData {
+        std::vector<T> A;
+        std::vector<T> Q;
+        std::vector<T> B;
+        std::vector<T> B_cpy;
+        std::vector<T> A_hat;
+        std::vector<T> A_k;
+        std::vector<T> A_cpy;
+        std::vector<T> A_cpy_2;
+        std::vector<T> A_cpy_3;
+        std::vector<T> s;
+        std::vector<T> S;
+        std::vector<T> U;
+        std::vector<T> VT;
+
+        QBTestData(int64_t m, int64_t n, int64_t k) :
+        A(m * n, 0.0), 
+        Q(m * k, 0.0), 
+        B(k * n, 0.0), 
+        B_cpy(k * n, 0.0), 
+        A_hat(m * n, 0.0),
+        A_k(m * n, 0.0),  
+        A_cpy(m * n, 0.0),  
+        A_cpy_2(m * n, 0.0),
+        A_cpy_3(m * n, 0.0),
+        s(n, 0.0),
+        S(n * n, 0.0),
+        U(m * n, 0.0),
+        VT(n * n, 0.0)
+        {}
+    };
+
+    template <typename T, typename RNG>
+    static void computational_helper(int64_t m, int64_t n, QBTestData<T>& all_data) {
+        
+        // Create a copy of the original matrix
+        blas::copy(m * n, all_data.A.data(), 1, all_data.A_cpy.data(), 1);
+        blas::copy(m * n, all_data.A.data(), 1, all_data.A_cpy_2.data(), 1);
+        blas::copy(m * n, all_data.A.data(), 1, all_data.A_cpy_3.data(), 1);
+
+        // Get low-rank SVD
+        lapack::gesdd(Job::SomeVec, m, n, all_data.A_cpy.data(), m, all_data.s.data(), all_data.U.data(), m, all_data.VT.data(), n);
+    }
+
+    template <typename T, typename RNG>
+    static void computational_helper(int64_t m, int64_t n, T& norm_A, QBTestData<T>& all_data) {
+
+        // pre-compute norm
+        norm_A = lapack::lange(Norm::Fro, m, n, all_data.A.data(), m);
+    }
```

Comment:
1. Don't use function names like this.
2. This function signature is awkward. Why not just return ``norm_A``?
3. This function is trivial. Just set ``all_data.norm_A = lapack::lange(...)`` elsewhere in the code, once you've populated ``all_data.A``.

Reply:
Once again, the only reason i added it is so that different test files have identical structure, but i can change that.

Reply:
Fixed.